functional-programming-spreadsheet/index980.js

```diff
@@ -0,0 +1,129 @@
+const infixToFunction = {
+  "+": (x, y) => x + y,
+  "-": (x, y) => x - y,
+  "*": (x, y) => x * y,
+  "/": (x, y) => x / y
+};
+
+const infixEval = (str, regex) =>
+  str.replace(regex, (_, arg1, fn, arg2) =>
+    infixToFunction[fn](parseFloat(arg1), parseFloat(arg2))
+  );
+
+const highPrecedence = str => {
+  const regex = /([0-9.]+)([*\/])([0-9.]+)/;
+  const str2 = infixEval(str, regex);
+  return str === str2 ? str : highPrecedence(str2);
+};
+
+const isEven = num => num % 2 === 0;
+const sum = nums => nums.reduce((a, x) => a + x);
+const average = nums => sum(nums) / nums.length;
+
+const median = nums => {
+  const sorted = nums.slice().sort((x, y) => x - y);
+  const length = sorted.length;
+  const middle = sorted.length / 2 - 1;
+  return isEven(length)
+    ? average([sorted[middle], sorted[middle + 1]])
+    : sorted[middle + 0.5];
+}; 
+
+const spreadsheetFunctions = {
+  "": x => x,
+  random: ([x, y]) => Math.floor(Math.random() * y + x),
+  increment: nums => nums.map(x => x + 1),
+  firsttwo: arr => arr.slice(0, 2),
+  lastttwo: arr => arr.slice(-2),
+  even: nums => nums.filter(isEven),
+  sum,
+  average,
+  median,
+  has2: arr => arr.includes(2),
+  nodups: arr => arr.reduce((a, x) => a.includes(x) ? a : a.concat(x), []),
+  range: arr => range(...arr),
+  someeven: arr => arr.some(isEven),
+  everyeven: arr => arr.every(isEven)
+};
+
+const applyFn = str => {
+  const noHigh = highPrecedence(str);
+  const infix = /([0-9.]+)([+-])([0-9.]+)/;
+  const str2 = infixEval(noHigh, infix);
+  const regex = /([a-z]*)\(([0-9., ]*)\)(?!.*\()/i;
+  const toNumberList = args => args.split(",").map(parseFloat);
+  const applyFunction = (fn, args) =>
+    spreadsheetFunctions[fn.toLowerCase()](toNumberList(args));
+  return str2.replace(
+    regex,
+    (match, fn, args) =>
+      spreadsheetFunctions.hasOwnProperty(fn.toLowerCase()) ? applyFunction(fn, args) : match
+  );
+};
+
+const range = (start, end) => Array(end - start + 1).fill(start);
+
+/*
+The function in the `map` method can actually take a second argument: the index of the element.
+This is why you need an arrow function in `charRange` - if you don't, then the index will be passed to `String.fromCharCode` as the second argument, leading to unexpected results.
+However, it is safe for functions like `parseFloat` which take only one argument (but not for `parseInt`).
+Use this form of `map` to add its index to every element in the array in `range`.
```

Comment:
I think it would be fine to tell learners exactly what to input here:

Chain `.map((x, i) => x + i)` to `.fill(start)` to add its index to every element in the array in `range`.